File: go/pkg/dashscope/realtime.go
package dashscope

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"iter"
	"net/http"
	"sync"

	"github.com/google/uuid"
	"github.com/gorilla/websocket"
)

// RealtimeService provides access to the Qwen-Omni-Realtime API.
type RealtimeService struct {
	client *Client
}

// Connect establishes a realtime session with the specified configuration.
// The session uses standard JSON messages similar to OpenAI Realtime API format.
func (s *RealtimeService) Connect(ctx context.Context, config *RealtimeConfig) (*RealtimeSession, error) {
	if config == nil {
		config = &RealtimeConfig{}
	}
	if config.Model == "" {
		config.Model = ModelQwenOmniTurboRealtimeLatest
	}

	// Build WebSocket URL: wss://dashscope.aliyuncs.com/api-ws/v1/realtime?model={model}
	url := fmt.Sprintf("%s?model=%s", s.client.config.baseURL, config.Model)

	// Build headers
	headers := http.Header{}
	headers.Set("Authorization", "bearer "+s.client.config.apiKey)
	if s.client.config.workspaceID != "" {
		headers.Set("X-DashScope-WorkSpace", s.client.config.workspaceID)
	}

	// Dial WebSocket
	dialer := websocket.Dialer{
		HandshakeTimeout: s.client.config.httpClient.Timeout,
	}

	conn, resp, err := dialer.DialContext(ctx, url, headers)
	if err != nil {
		if resp != nil {
			return nil, &Error{
				Code:       "ConnectionFailed",
				Message:    fmt.Sprintf("failed to connect: %v", err),
				HTTPStatus: resp.StatusCode,
			}
		}
		return nil, fmt.Errorf("dashscope: failed to connect: %w", err)
	}

	session := &RealtimeSession{
		conn:     conn,
		config:   config,
		client:   s.client,
		closeCh:  make(chan struct{}),
		eventsCh: make(chan eventOrError, 100),
	}

	// Start background reader
	go session.readLoop()

	return session, nil
}

// RealtimeSession represents an active realtime session.
type RealtimeSession struct {
	conn      *websocket.Conn
	config    *RealtimeConfig
	client    *Client
	sessionID string
	closeCh   chan struct{}
	eventsCh  chan eventOrError
	closeOnce sync.Once
	mu        sync.Mutex
}

type eventOrError struct {
	event *RealtimeEvent
	err   error
}

// generateEventID generates a unique event ID.
func generateEventID() string {
	return "event_" + uuid.New().String()[:8]
}

// UpdateSession updates the session configuration.
// This should be called after receiving session.created event.
func (s *RealtimeSession) UpdateSession(config *SessionConfig) error {
	sessionConfig := map[string]interface{}{}

	if len(config.Modalities) > 0 {
		sessionConfig["modalities"] = config.Modalities
	}
	if config.Voice != "" {
		sessionConfig["voice"] = config.Voice
	}
	if config.InputAudioFormat != "" {
		sessionConfig["input_audio_format"] = config.InputAudioFormat
	}
	if config.OutputAudioFormat != "" {
		sessionConfig["output_audio_format"] = config.OutputAudioFormat
	}
	if config.Instructions != "" {
		sessionConfig["instructions"] = config.Instructions
	}
	if config.EnableInputAudioTranscription {
		// Only send if model is specified, otherwise server uses default
		if config.InputAudioTranscriptionModel != "" {
			sessionConfig["input_audio_transcription"] = map[string]interface{}{
				"model": config.InputAudioTranscriptionModel,
			}
		}
	}
	if config.TurnDetection != nil {
		turnDetection := map[string]interface{}{
			"type": config.TurnDetection.Type,
		}
		if config.TurnDetection.Threshold > 0 {
			turnDetection["threshold"] = config.TurnDetection.Threshold
		}
		if config.TurnDetection.PrefixPaddingMs > 0 {
			turnDetection["prefix_padding_ms"] = config.TurnDetection.PrefixPaddingMs
		}
		if config.TurnDetection.SilenceDurationMs > 0 {
			turnDetection["silence_duration_ms"] = config.TurnDetection.SilenceDurationMs
		}
		sessionConfig["turn_detection"] = turnDetection
	}

	return s.sendEvent(map[string]interface{}{
		"event_id": generateEventID(),
		"type":     "session.update",
		"session":  sessionConfig,
	})
}

// AppendAudio sends audio data to the input audio buffer.
// Audio should be base64 encoded PCM data.
func (s *RealtimeSession) AppendAudio(audio []byte) error {
	encoded := base64.StdEncoding.EncodeToString(audio)
	return s.sendEvent(map[string]interface{}{
		"event_id": generateEventID(),
		"type":     "input_audio_buffer.append",
		"audio":    encoded,
	})
}

// AppendAudioBase64 sends base64-encoded audio data to the input audio buffer.
func (s *RealtimeSession) AppendAudioBase64(audioBase64 string) error {
	return s.sendEvent(map[string]interface{}{
		"event_id": generateEventID(),
		"type":     "input_audio_buffer.append",
		"audio":    audioBase64,
	})
}

// AppendImage sends an image frame for video input.
// Image should be base64 encoded.
func (s *RealtimeSession) AppendImage(image []byte) error {
	encoded := base64.StdEncoding.EncodeToString(image)
	return s.sendEvent(map[string]interface{}{
		"event_id": generateEventID(),
		"type":     "input_image_buffer.append",
		"image":    encoded,
	})
}

// CommitInput commits the audio buffer.
// In server_vad mode, this is called automatically after VAD detects end of speech.
// In manual mode, call this to indicate end of user input.
func (s *RealtimeSession) CommitInput() error {
	return s.sendEvent(map[string]interface{}{
		"event_id": generateEventID(),
		"type":     "input_audio_buffer.commit",
	})
}

// ClearInput clears the input audio buffer.
func (s *RealtimeSession) ClearInput() error {
	return s.sendEvent(map[string]interface{}{
		"event_id": generateEventID(),
		"type":     "input_audio_buffer.clear",
	})
}

// CreateResponse requests the model to generate a response.
// In server_vad mode, this is called automatically by the server.
// In manual mode, call this after CommitInput to trigger response generation.
func (s *RealtimeSession) CreateResponse(opts *ResponseCreateOptions) error {
	event := map[string]interface{}{
		"event_id": generateEventID(),
		"type":     "response.create",
		"response": map[string]interface{}{},
	}

	if opts != nil {
		response := event["response"].(map[string]interface{})
		if opts.Instructions != "" {
			response["instructions"] = opts.Instructions
		}
		if len(opts.Modalities) > 0 {
			response["modalities"] = opts.Modalities
		}
	}

	return s.sendEvent(event)
}

// CancelResponse cancels the current response generation.
func (s *RealtimeSession) CancelResponse() error {
	return s.sendEvent(map[string]interface{}{
		"event_id": generateEventID(),
		"type":     "response.cancel",
	})
}

// FinishSession sends a session.finish event to gracefully end the session.
func (s *RealtimeSession) FinishSession() error {
	return s.sendEvent(map[string]interface{}{
		"event_id": generateEventID(),
		"type":     "session.finish",
	})
}

// SendRaw sends a raw JSON event to the server.
// Use this for events not covered by helper methods.
func (s *RealtimeSession) SendRaw(event map[string]interface{}) error {
	return s.sendEvent(event)
}

// Events returns an iterator over session events.
func (s *RealtimeSession) Events() iter.Seq2[*RealtimeEvent, error] {
	return func(yield func(*RealtimeEvent, error) bool) {
		for {
			select {
			case <-s.closeCh:
				return
			case item, ok := <-s.eventsCh:
				if !ok {
					return
				}
				if !yield(item.event, item.err) {
					return
				}
				if item.err != nil {
					return
				}
			}
		}
	}
}

// Close closes the session.
func (s *RealtimeSession) Close() error {
	var err error
	s.closeOnce.Do(func() {
		close(s.closeCh)
		err = s.conn.Close()
	})
	return err
}

// SessionID returns the session ID assigned by the server.
func (s *RealtimeSession) SessionID() string {
	return s.sessionID
}

// sendEvent sends a JSON event to the server.
func (s *RealtimeSession) sendEvent(event map[string]interface{}) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Debug: log the event being sent
	if s.client.config.debug {
		if jsonBytes, err := json.MarshalIndent(event, "", "  "); err == nil {
			// Truncate for readability
			str := string(jsonBytes)
			if len(str) > 500 {
				str = str[:500] + "..."
			}
			fmt.Printf("[DEBUG] Sending: %s\n", str)
		}
	}

	return s.conn.WriteJSON(event)
}

// readLoop reads events from the WebSocket connection.
func (s *RealtimeSession) readLoop() {
	defer close(s.eventsCh)

	for {
		select {
		case <-s.closeCh:
			return
		default:
		}

		_, message, err := s.conn.ReadMessage()
		if err != nil {
			select {
			case <-s.closeCh:
				return
			case s.eventsCh <- eventOrError{err: fmt.Errorf("read error: %w", err)}:
			}
			return
		}

		// Debug: log received message
		if s.client.config.debug {
			msgStr := string(message)
			if len(msgStr) > 1000 {
				msgStr = msgStr[:1000] + "..."
			}
			fmt.Printf("[DEBUG] Received (len=%d): %s\n", len(message), msgStr)
		}

		// Parse JSON event
		var rawEvent map[string]json.RawMessage
		if err := json.Unmarshal(message, &rawEvent); err != nil {
			select {
			case <-s.closeCh:
				return
			case s.eventsCh <- eventOrError{err: fmt.Errorf("parse error: %w", err)}:
			}
			continue
		}

		// Extract event type
		var eventType string
		if typeRaw, ok := rawEvent["type"]; ok {
			json.Unmarshal(typeRaw, &eventType)
		}

		// Check for error event
		if eventType == "error" {
			var errorData struct {
				Type  string `json:"type"`
				Error struct {
					Type    string `json:"type"`
					Code    string `json:"code"`
					Message string `json:"message"`
				} `json:"error"`
			}
			if err := json.Unmarshal(message, &errorData); err == nil {
				select {
				case <-s.closeCh:
					return
				case s.eventsCh <- eventOrError{err: &Error{
					Code:    errorData.Error.Code,
					Message: errorData.Error.Message,
				}}:
				}
				continue
			}
		}

		// Convert to RealtimeEvent
		event := s.parseEvent(eventType, message)
		if event != nil {
			// Track session ID
			if eventType == "session.created" && event.Session != nil {
				s.sessionID = event.Session.ID
			}

			select {
			case <-s.closeCh:
				return
			case s.eventsCh <- eventOrError{event: event}:
			}
		}
	}
}

// parseEvent parses a raw JSON message into a RealtimeEvent.
func (s *RealtimeSession) parseEvent(eventType string, message []byte) *RealtimeEvent {
	event := &RealtimeEvent{
		Type: eventType,
	}

	// Check for DashScope "choices" format (different from OpenAI Realtime events)
	var choicesData struct {
		Choices []struct {
			FinishReason string `json:"finish_reason"`
			Message      struct {
				Role    string `json:"role"`
				Content []struct {
					Text  string `json:"text,omitempty"`
					Audio *struct {
						Data string `json:"data"`
					} `json:"audio,omitempty"`
				} `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.Unmarshal(message, &choicesData); err == nil && len(choicesData.Choices) > 0 {
		// DashScope "choices" format response
		event.Type = EventTypeChoicesResponse
		choice := choicesData.Choices[0]

		for _, content := range choice.Message.Content {
			if content.Text != "" {
				event.Delta = content.Text
			}
			if content.Audio != nil && content.Audio.Data != "" {
				event.AudioBase64 = content.Audio.Data
				if decoded, err := base64.StdEncoding.DecodeString(content.Audio.Data); err == nil {
					event.Audio = decoded
				}
			}
		}

		if choice.FinishReason != "" && choice.FinishReason != "null" {
			event.FinishReason = choice.FinishReason
		}

		return event
	}

	// Standard event format
	switch eventType {
	case EventTypeSessionCreated, EventTypeSessionUpdated:
		var data struct {
			Session struct {
				ID string `json:"id"`
			} `json:"session"`
		}
		if err := json.Unmarshal(message, &data); err == nil && data.Session.ID != "" {
			event.Session = &SessionInfo{ID: data.Session.ID}
		}

	case EventTypeResponseCreated:
		var data struct {
			Response struct {
				ID string `json:"id"`
			} `json:"response"`
		}
		if err := json.Unmarshal(message, &data); err == nil {
			event.ResponseID = data.Response.ID
		}

	case EventTypeResponseAudioDelta:
		var data struct {
			Delta string `json:"delta"`
		}
		if err := json.Unmarshal(message, &data); err == nil && data.Delta != "" {
			event.AudioBase64 = data.Delta
			if decoded, err := base64.StdEncoding.DecodeString(data.Delta); err == nil {
				event.Audio = decoded
			}
		}

	case EventTypeResponseTranscriptDelta:
		var data struct {
			Delta string `json:"delta"`
		}
		if err := json.Unmarshal(message, &data); err == nil {
			event.Delta = data.Delta
		}

	case EventTypeResponseTextDelta:
		var data struct {
			Delta string `json:"delta"`
		}
		if err := json.Unmarshal(message, &data); err == nil {
			event.Delta = data.Delta
		}

	case "conversation.item.input_audio_transcription.completed":
		var data struct {
			Transcript string `json:"transcript"`
		}
		if err := json.Unmarshal(message, &data); err == nil {
			event.Transcript = data.Transcript
		}

	case EventTypeResponseDone:
		var data struct {
			Response struct {
				Usage struct {
					TotalTokens  int `json:"total_tokens"`
					InputTokens  int `json:"input_tokens"`
					OutputTokens int `json:"output_tokens"`
				} `json:"usage"`
			} `json:"response"`
		}
		if err := json.Unmarshal(message, &data); err == nil {
			event.Usage = &UsageStats{
				TotalTokens:  data.Response.Usage.TotalTokens,
				InputTokens:  data.Response.Usage.InputTokens,
				OutputTokens: data.Response.Usage.OutputTokens,
			}
		}
	}

	return event
}

// ResponseCreateOptions contains options for creating a response.
type ResponseCreateOptions struct {
	// Instructions override for this response (optional)
	Instructions string
	// Output modalities for this response (optional)
	Modalities []string
}

File: rust/dashscope/src/realtime.rs
//! Realtime WebSocket API for Qwen-Omni-Realtime.

use std::sync::Arc;

use base64::Engine;
use futures::stream::{Stream, StreamExt};
use serde::{Deserialize, Serialize};
use serde_json::json;
use tokio::sync::mpsc;
use tokio_tungstenite::{connect_async, tungstenite::Message};
use tracing::{debug, error};

use crate::{
    client::ClientConfig,
    error::{Error, Result},
    event::*,
    types::*,
    MODEL_QWEN_OMNI_TURBO_REALTIME_LATEST,
};

/// Realtime service for Qwen-Omni-Realtime API.
pub struct RealtimeService {
    config: Arc<ClientConfig>,
}

impl RealtimeService {
    pub(crate) fn new(config: Arc<ClientConfig>) -> Self {
        Self { config }
    }

    /// Connects to a realtime session.
    pub async fn connect(&self, realtime_config: &RealtimeConfig) -> Result<RealtimeSession> {
        let model = if realtime_config.model.is_empty() {
            MODEL_QWEN_OMNI_TURBO_REALTIME_LATEST
        } else {
            &realtime_config.model
        };

        // Build WebSocket URL
        let url = format!("{}?model={}", self.config.base_url, model);
        debug!("Connecting to: {}", url);

        // Build request with headers
        let request = http::Request::builder()
            .uri(&url)
            .header("Authorization", format!("bearer {}", self.config.api_key));

        let request = if let Some(ref workspace_id) = self.config.workspace_id {
            request.header("X-DashScope-WorkSpace", workspace_id)
        } else {
            request
        };

        let request = request
            .header("Sec-WebSocket-Version", "13")
            .header("Sec-WebSocket-Key", generate_websocket_key())
            .header("Connection", "Upgrade")
            .header("Upgrade", "websocket")
            .header("Host", extract_host(&url).unwrap_or("dashscope.aliyuncs.com"))
            .body(())
            .map_err(|e| Error::Connection(format!("Failed to build request: {}", e)))?;

        // Connect
        let (ws_stream, _response) = connect_async(request)
            .await
            .map_err(|e| Error::Connection(format!("Failed to connect: {}", e)))?;

        let (write, read) = ws_stream.split();

        // Create channels
        let (event_tx, event_rx) = mpsc::channel(100);
        let (write_tx, write_rx) = mpsc::channel(100);

        // Spawn write task
        let write_handle = tokio::spawn(write_loop(write, write_rx));

        // Spawn read task
        let event_tx_clone = event_tx.clone();
        let read_handle = tokio::spawn(read_loop(read, event_tx_clone));

        Ok(RealtimeSession {
            config: realtime_config.clone(),
            event_rx,
            write_tx,
            _read_handle: read_handle,
            _write_handle: write_handle,
            session_id: None,
        })
    }
}

/// Realtime session configuration.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct RealtimeConfig {
    /// Model ID to use. Default: qwen-omni-turbo-realtime-latest
    #[serde(default)]
    pub model: String,
}

/// Options for creating a response.
#[derive(Debug, Clone, Default)]
pub struct ResponseCreateOptions {
    /// Messages for text input (DashScope-specific).
    pub messages: Option<Vec<SimpleMessage>>,
    /// Instructions override for this response.
    pub instructions: Option<String>,
    /// Output modalities for this response.
    pub modalities: Option<Vec<String>>,
}

/// Simple message with role and content.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimpleMessage {
    pub role: String,
    pub content: String,
}

impl SimpleMessage {
    /// Creates a new text message.
    pub fn new(role: impl Into<String>, content: impl Into<String>) -> Self {
        Self {
            role: role.into(),
            content: content.into(),
        }
    }

    /// Creates a user message.
    pub fn user(content: impl Into<String>) -> Self {
        Self::new("user", content)
    }

    /// Creates an assistant message.
    pub fn assistant(content: impl Into<String>) -> Self {
        Self::new("assistant", content)
    }
}

/// Active realtime session.
pub struct RealtimeSession {
    config: RealtimeConfig,
    event_rx: mpsc::Receiver<Result<RealtimeEvent>>,
    write_tx: mpsc::Sender<Message>,
    _read_handle: tokio::task::JoinHandle<()>,
    _write_handle: tokio::task::JoinHandle<()>,
    session_id: Option<String>,
}

impl RealtimeSession {
    /// Returns the session configuration.
    pub fn config(&self) -> &RealtimeConfig {
        &self.config
    }

    /// Returns the session ID (if available).
    pub fn session_id(&self) -> Option<&str> {
        self.session_id.as_deref()
    }

    /// Updates the session configuration.
    pub async fn update_session(&self, config: &SessionConfig) -> Result<()> {
        let mut session_config = serde_json::Map::new();

        if let Some(ref modalities) = config.modalities {
            session_config.insert("modalities".to_string(), json!(modalities));
        }
        if let Some(ref voice) = config.voice {
            session_config.insert("voice".to_string(), json!(voice));
        }
        if let Some(ref format) = config.input_audio_format {
            session_config.insert("input_audio_format".to_string(), json!(format));
        }
        if let Some(ref format) = config.output_audio_format {
            session_config.insert("output_audio_format".to_string(), json!(format));
        }
        if let Some(ref instructions) = config.instructions {
            session_config.insert("instructions".to_string(), json!(instructions));
        }
        if let Some(enable) = config.enable_input_audio_transcription {
            if enable {
                let mut transcription = serde_json::Map::new();
                if let Some(ref model) = config.input_audio_transcription_model {
                    transcription.insert("model".to_string(), json!(model));
                }
                session_config.insert("input_audio_transcription".to_string(), json!(transcription));
            }
        }
        if let Some(ref td) = config.turn_detection {
            let mut turn_detection = serde_json::Map::new();
            if let Some(ref t) = td.detection_type {
                turn_detection.insert("type".to_string(), json!(t));
            }
            if let Some(threshold) = td.threshold {
                turn_detection.insert("threshold".to_string(), json!(threshold));
            }
            if let Some(prefix) = td.prefix_padding_ms {
                turn_detection.insert("prefix_padding_ms".to_string(), json!(prefix));
            }
            if let Some(silence) = td.silence_duration_ms {
                turn_detection.insert("silence_duration_ms".to_string(), json!(silence));
            }
            session_config.insert("turn_detection".to_string(), json!(turn_detection));
        }

        let event = json!({
            "event_id": generate_event_id(),
            "type": EVENT_TYPE_SESSION_UPDATE,
            "session": session_config,
        });

        self.send_event(event).await
    }

    /// Appends audio data to the input buffer.
    pub async fn append_audio(&self, audio: &[u8]) -> Result<()> {
        let encoded = base64::engine::general_purpose::STANDARD.encode(audio);
        let event = json!({
            "event_id": generate_event_id(),
            "type": EVENT_TYPE_INPUT_AUDIO_APPEND,
            "audio": encoded,
        });
        self.send_event(event).await
    }

    /// Appends base64-encoded audio data to the input buffer.
    pub async fn append_audio_base64(&self, audio_base64: &str) -> Result<()> {
        let event = json!({
            "event_id": generate_event_id(),
            "type": EVENT_TYPE_INPUT_AUDIO_APPEND,
            "audio": audio_base64,
        });
        self.send_event(event).await
    }

    /// Appends an image frame to the input buffer.
    pub async fn append_image(&self, image: &[u8]) -> Result<()> {
        let encoded = base64::engine::general_purpose::STANDARD.encode(image);
        let event = json!({
            "event_id": generate_event_id(),
            "type": EVENT_TYPE_INPUT_IMAGE_APPEND,
            "image": encoded,
        });
        self.send_event(event).await
    }

    /// Commits the input audio buffer.
    pub async fn commit_input(&self) -> Result<()> {
        let event = json!({
            "event_id": generate_event_id(),
            "type": EVENT_TYPE_INPUT_AUDIO_COMMIT,
        });
        self.send_event(event).await
    }

    /// Clears the input audio buffer.
    pub async fn clear_input(&self) -> Result<()> {
        let event = json!({
            "event_id": generate_event_id(),
            "type": EVENT_TYPE_INPUT_AUDIO_CLEAR,
        });
        self.send_event(event).await
    }

    /// Creates a response.
    pub async fn create_response(&self, opts: Option<&ResponseCreateOptions>) -> Result<()> {
        let mut event = json!({
            "event_id": generate_event_id(),
            "type": EVENT_TYPE_RESPONSE_CREATE,
        });

        if let Some(opts) = opts {
            if let Some(ref messages) = opts.messages {
                event["messages"] = json!(messages);
            }

            let mut response = serde_json::Map::new();
            if let Some(ref instructions) = opts.instructions {
                response.insert("instructions".to_string(), json!(instructions));
            }
            if let Some(ref modalities) = opts.modalities {
                response.insert("modalities".to_string(), json!(modalities));
            }
            if !response.is_empty() {
                event["response"] = json!(response);
            }
        }

        self.send_event(event).await
    }

    /// Cancels the current response.
    pub async fn cancel_response(&self) -> Result<()> {
        let event = json!({
            "event_id": generate_event_id(),
            "type": EVENT_TYPE_RESPONSE_CANCEL,
        });
        self.send_event(event).await
    }

    /// Finishes the session gracefully.
    pub async fn finish_session(&self) -> Result<()> {
        let event = json!({
            "event_id": generate_event_id(),
            "type": EVENT_TYPE_SESSION_FINISH,
        });
        self.send_event(event).await
    }

    /// Sends a raw JSON event.
    pub async fn send_raw(&self, event: serde_json::Value) -> Result<()> {
        self.send_event(event).await
    }

    /// Returns a stream of events.
    pub fn events(&mut self) -> impl Stream<Item = Result<RealtimeEvent>> + '_ {
        futures::stream::poll_fn(move |cx| {
            self.event_rx.poll_recv(cx).map(|opt| opt)
        })
    }

    /// Receives the next event.
    pub async fn recv(&mut self) -> Option<Result<RealtimeEvent>> {
        self.event_rx.recv().await
    }

    /// Closes the session.
    pub async fn close(&self) -> Result<()> {
        let _ = self.write_tx.send(Message::Close(None)).await;
        Ok(())
    }

    async fn send_event(&self, event: serde_json::Value) -> Result<()> {
        let msg = Message::Text(event.to_string().into());
        debug!("Sending event: {}", truncate_for_log(&event.to_string(), 500));
        self.write_tx
            .send(msg)
            .await
            .map_err(|_| Error::SessionClosed)
    }
}

// Write loop task
async fn write_loop(
    mut write: futures::stream::SplitSink<
        tokio_tungstenite::WebSocketStream<
            tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>,
        >,
        Message,
    >,
    mut rx: mpsc::Receiver<Message>,
) {
    use futures::SinkExt;

    while let Some(msg) = rx.recv().await {
        if let Message::Close(_) = msg {
            let _ = write.close().await;
            break;
        }
        if let Err(e) = write.send(msg).await {
            error!("Write error: {}", e);
            break;
        }
    }
}

// Read loop task
async fn read_loop(
    mut read: futures::stream::SplitStream<
        tokio_tungstenite::WebSocketStream<
            tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>,
        >,
    >,
    tx: mpsc::Sender<Result<RealtimeEvent>>,
) {
    while let Some(result) = read.next().await {
        match result {
            Ok(Message::Text(text)) => {
                debug!("Received: {}", truncate_for_log(&text, 1000));
                let event = parse_event(&text);
                if tx.send(event).await.is_err() {
                    break;
                }
            }
            Ok(Message::Close(_)) => {
                debug!("WebSocket closed by server");
                break;
            }
            Ok(Message::Ping(data)) => {
                debug!("Received ping: {:?}", data);
            }
            Ok(_) => {}
            Err(e) => {
                error!("Read error: {}", e);
                let _ = tx.send(Err(Error::WebSocket(e))).await;
                break;
            }
        }
    }
}

// Parse event from JSON
fn parse_event(text: &str) -> Result<RealtimeEvent> {
    let raw: serde_json::Value = serde_json::from_str(text)?;
    let mut event = RealtimeEvent::default();

    // Extract event type
    if let Some(t) = raw.get("type").and_then(|v| v.as_str()) {
        event.event_type = t.to_string();
    }

    // Extract event ID
    if let Some(id) = raw.get("event_id").and_then(|v| v.as_str()) {
        event.event_id = Some(id.to_string());
    }

    // Check for DashScope "choices" format
    if let Some(choices) = raw.get("choices").and_then(|v| v.as_array()) {
        event.event_type = EVENT_TYPE_CHOICES_RESPONSE.to_string();
        
        if let Some(choice) = choices.first() {
            // Extract finish_reason
            if let Some(reason) = choice.get("finish_reason").and_then(|v| v.as_str()) {
                if reason != "null" {
                    event.finish_reason = Some(reason.to_string());
                }
            }

            // Extract message content
            if let Some(message) = choice.get("message") {
                if let Some(content) = message.get("content").and_then(|v| v.as_array()) {
                    for item in content {
                        // Text content
                        if let Some(text) = item.get("text").and_then(|v| v.as_str()) {
                            event.delta = Some(text.to_string());
                        }
                        // Audio content
                        if let Some(audio) = item.get("audio") {
                            if let Some(data) = audio.get("data").and_then(|v| v.as_str()) {
                                event.audio_base64 = Some(data.to_string());
                                if let Ok(decoded) = base64::engine::general_purpose::STANDARD.decode(data) {
                                    event.audio = Some(decoded);
                                }
                            }
                        }
                    }
                }
            }
        }
        return Ok(event);
    }

    // Standard event format parsing
    match event.event_type.as_str() {
        EVENT_TYPE_SESSION_CREATED | EVENT_TYPE_SESSION_UPDATED => {
            if let Some(session) = raw.get("session") {
                if let Ok(info) = serde_json::from_value(session.clone()) {
                    event.session = Some(info);
                }
            }
        }
        EVENT_TYPE_RESPONSE_CREATED => {
            if let Some(response) = raw.get("response") {
                if let Some(id) = response.get("id").and_then(|v| v.as_str()) {
                    event.response_id = Some(id.to_string());
                }
            }
        }
        EVENT_TYPE_RESPONSE_AUDIO_DELTA => {
            if let Some(delta) = raw.get("delta").and_then(|v| v.as_str()) {
                event.audio_base64 = Some(delta.to_string());
                if let Ok(decoded) = base64::engine::general_purpose::STANDARD.decode(delta) {
                    event.audio = Some(decoded);
                }
            }
        }
        EVENT_TYPE_RESPONSE_TRANSCRIPT_DELTA | EVENT_TYPE_RESPONSE_TEXT_DELTA => {
            if let Some(delta) = raw.get("delta").and_then(|v| v.as_str()) {
                event.delta = Some(delta.to_string());
            }
        }
        EVENT_TYPE_INPUT_AUDIO_TRANSCRIPTION_COMPLETED => {
            if let Some(transcript) = raw.get("transcript").and_then(|v| v.as_str()) {
                event.transcript = Some(transcript.to_string());
            }
        }
        EVENT_TYPE_RESPONSE_DONE => {
            if let Some(response) = raw.get("response") {
                if let Ok(info) = serde_json::from_value::<ResponseInfo>(response.clone()) {
                    event.response = Some(info.clone());
                    event.usage = info.usage;
                }
            }
        }
        EVENT_TYPE_ERROR => {
            if let Some(error) = raw.get("error") {
                if let Ok(err) = serde_json::from_value(error.clone()) {
                    event.error = Some(err);
                }
            }
        }
        _ => {}
    }

    Ok(event)
}

// Helper functions
fn generate_event_id() -> String {
    format!("event_{}", &uuid::Uuid::new_v4().to_string()[..12])
}

fn generate_websocket_key() -> String {
    use base64::Engine;
    let mut key = [0u8; 16];
    for byte in &mut key {
        *byte = rand_byte();
    }
    base64::engine::general_purpose::STANDARD.encode(key)
}

fn rand_byte() -> u8 {
    use std::time::{SystemTime, UNIX_EPOCH};
    let duration = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    ((duration.as_nanos() % 256) as u8).wrapping_add(duration.subsec_nanos() as u8)
}

fn extract_host(url: &str) -> Option<&str> {
    url.strip_prefix("wss://")
        .or_else(|| url.strip_prefix("ws://"))
        .and_then(|s| s.split('/').next())
        .and_then(|s| s.split('?').next())
}

fn truncate_for_log(s: &str, max_len: usize) -> String {
    if s.len() > max_len {
        format!("{}...", &s[..max_len])
    } else {
        s.to_string()
    }
}
